shared/src/balancer/pool_init.rs

```diff
@@ -0,0 +1,256 @@
+//! Balancer pool registry initialization.
+//!
+//! This module contains a component used to initialize Balancer pool registries
+//! with existing data in order to reduce the "cold start" time of the service.
+
+use crate::balancer::{
+    graph_api::{BalancerSubgraphClient, RegisteredWeightedPools},
+    pool_storage::RegisteredWeightedPool,
+};
+use anyhow::{anyhow, Result};
+use contracts::{BalancerV2WeightedPool2TokensFactory, BalancerV2WeightedPoolFactory};
+use ethcontract::{Artifact, H160};
+
+#[derive(Debug, Default, PartialEq)]
+pub struct BalancerRegisteredPools {
+    pub weighted_pools: Vec<RegisteredWeightedPool>,
+    pub weighted_2token_pools: Vec<RegisteredWeightedPool>,
+}
+
+#[async_trait::async_trait]
+pub trait PoolInitializing: Send + Sync {
+    async fn initialize_pools(&self) -> Result<BalancerRegisteredPools>;
+}
+
+/// A Balancer pool registry initializer that always returns empty pools.
+///
+/// This can be used to index all pools from events instead of relying on the
+/// Balancer subgraph for example.
+pub struct EmptyPoolInitializer;
+
+#[async_trait::async_trait]
+impl PoolInitializing for EmptyPoolInitializer {
+    async fn initialize_pools(&self) -> Result<BalancerRegisteredPools> {
+        Ok(Default::default())
+    }
+}
```

Comment:
Doesn't it seem like overkill to make traits for this if we know we will only ever need it once? If it were going to be mocked in the test framework, then I see the point, but that doesn't seem to be the case.

Reply:
This allows mocking and testing of how it interacts with the `BalancerPoolFetcher` once it gets hooked up. Additionally, it allows us to also have a `FilePoolInitializer` and `DatabasePoolInitializer` if we ever want to move away from the graph without having to refactor the calling code.